6. Хеш-таблица

VitaliyIliyushenkov opened the pull request.

VitaliyIliyushenkov committed:
new file:   hash_tabl/hash_tabl.sln
	new file:   hash_tabl/hash_tabl/App.config
	new file:   hash_tabl/hash_tabl/Form1.Designer.cs
	new file:   hash_tabl/hash_tabl/Form1.cs
	new file:   hash_tabl/hash_tabl/Form1.resx
	new file:   hash_tabl/hash_tabl/HashItem.cs
	new file:   hash_tabl/hash_tabl/MyHashTable.cs
	new file:   hash_tabl/hash_tabl/Program.cs
	new file:   hash_tabl/hash_tabl/Properties/AssemblyInfo.cs
	new file:   hash_tabl/hash_tabl/Properties/Resources.Designer.cs
	new file:   hash_tabl/hash_tabl/Properties/Resources.resx
	new file:   hash_tabl/hash_tabl/Properties/Settings.Designer.cs
	new file:   hash_tabl/hash_tabl/Properties/Settings.settings
	new file:   hash_tabl/hash_tabl/hash_tabl.csproj

VitaliyIliyushenkov committed:
Delete queue)app directory

VitaliyIliyushenkov committed:
Delete stack_app directory

VasilevsWork commented on hash_tabl/hash_tabl/MyHashTable.cs:
```diff
@@ -0,0 +1,127 @@
+{
+    internal class MyHashTable<Tkey, TValue> : IEnumerable<HashItem<Tkey, TValue>>
+    {
+        HashItem<Tkey, TValue>[] hash_mass = new HashItem<Tkey, TValue>[0];
```
Лучше опираться список внутри массива. Не требуется реализовывать цепочки самостоятельно.

VitaliyIliyushenkov replied:
Теперь массив хранит класс List:
List<KeyValuePair<Tkey, TValue>>[] hash_mass = new List<KeyValuePair<Tkey, TValue>>[10000];

VasilevsWork commented on hash_tabl/hash_tabl/MyHashTable.cs:
```diff
@@ -0,0 +1,127 @@
+            HashItem<Tkey, TValue> newItem = new HashItem<Tkey, TValue> { Key = key, Value = value, Hash = hash};
+
+
+            if (hash >= hash_mass.Length)
+                Array.Resize(ref hash_mass, hash+1);
```
Значение хеша не должно определять размер массива. Массив по размеру фиксированный. Значения хеша нормируются.

VitaliyIliyushenkov replied:
Массив теперь фиксированного размера - 10 000
Хеш вычисляется и нормализируется по формуле:
int hash = Math.Abs(key.GetHashCode()) % 10000;

VasilevsWork commented on hash_tabl/hash_tabl/MyHashTable.cs:
```diff
@@ -0,0 +1,127 @@
+        }
+        public void Delete(Tkey key) 
+        {
+            int hash = HashFunction(key);
```
Используйте список и воспользуйтесь встроенными методами.

VitaliyIliyushenkov replied:
Удаляю последний элемент в листе
public void Delete(Tkey key)
{
    int hash = Math.Abs(key.GetHashCode()) % 10000;

    if (hash_mass[hash] == null)
        throw new Exception("Эл-т не найден");
    hash_mass[hash].RemoveAt(hash_mass[hash].Count-1);
}

VitaliyIliyushenkov replied:
Использовал функции листа, такие как Remove():

public KeyValuePair<Tkey, TValue> Find(Tkey key)
{
    int hash = Math.Abs(key.GetHashCode()) % 10000;
    if (hash_mass[hash] == null)
        throw new Exception("Эл-т не найден");
    foreach (var item in hash_mass[hash])
    {
        if(item.Key.Equals(key))
            return item;
    }
    throw new Exception("Эл-т не найден");
}

public void Delete(Tkey key)
{
    int hash = Math.Abs(key.GetHashCode()) % 10000;
    hash_mass[hash].Remove(Find(key));
}

VasilevsWork commented on hash_tabl/hash_tabl/MyHashTable.cs:
```diff
@@ -0,0 +1,127 @@
+            return list;
+        }
+
+        public IEnumerator<HashItem<Tkey, TValue>> GetEnumerator()
```
Нужно реализовать собственный `Enumerator`. Он должен самостоятельно посещать элементы хеш-таблицы.

VitaliyIliyushenkov replied:
Добавил собственный класс MyHashEnum, который преобразует хеш-таблицу в линейный лист и считывает каждый эл-т

VasilevsWork reviewed (commented):
Задание проверено на семинаре. Общее направление реализации верное, однако требуется ряд доработок для соответствия ожидаемому результату.

VitaliyIliyushenkov committed:
modified:   hash_tabl/hash_tabl/Form1.Designer.cs
	modified:   hash_tabl/hash_tabl/Form1.cs
	new file:   hash_tabl/hash_tabl/MyHashEnum.cs
	modified:   hash_tabl/hash_tabl/MyHashTable.cs
	modified:   hash_tabl/hash_tabl/hash_tabl.csproj

VitaliyIliyushenkov committed:
modified:   hash_tabl/hash_tabl/Form1.Designer.cs
	modified:   hash_tabl/hash_tabl/Form1.cs
	modified:   hash_tabl/hash_tabl/MyHashTable.cs

VasilevsWork reviewed (approved):
Задание принято.
